components/arm.py

```diff
@@ -0,0 +1,216 @@
+
+
+class Arm:
+    # masses used to calculate gravity feedforwards
+    ARM_END_MASS = 5  # kg
+    ARM_BASE_MASS = 1
+    ROTATE_GRAVITY_FEEDFORWARD = 1
+    EXTEND_GRAVITY_FEEDFORWARD = 1
+
+    MIN_EXTENSION = 0.7  # meters
+    MAX_EXTENSION = 1.3
+
+    ROTATE_GEAR_RATIO = (60 / 25) * (60 / 25) * (70 / 20)
+    EXTEND_GEAR_RATIO = (1 / 7) * (math.pi * 0.05)  # converts to meters
+
+    ANGLE_BOUNDARIES = (
+        90,
+        270,
+    )  # a range of angles that the arm cannot turn to. in the case of (90, 270), the arm wouldn't be able to move to 90 degrees, 120 degrees, etc. to 270 degrees
+    # TODO: add the real boundaries
+
+    def __init__(self):
+
+        # Create rotation things
+        # left motor is main
+        self.rotation_motor = rev.CANSparkMax(
+            CanIds.Arm.rotation_left, rev.CANSparkMax.MotorType.kBrushless
+        )
+        self.rotation_encoder = self.rotation_motor.getEncoder()
+        self._rotation_motor_right = rev.CANSparkMax(
+            CanIds.Arm.rotation_right, rev.CANSparkMax.MotorType.kBrushless
+        )
+        self._rotation_motor_right.follow(self.rotation_motor, True)
+        self._rotation_motor_right.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
+        self.rotation_motor.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
+        # output position = conversion factor * motor rotations
+        self.rotation_encoder.setPositionConversionFactor(self.ROTATE_GEAR_RATIO)
+        # also go from RPM to RPS
+        self.rotation_encoder.setVelocityConversionFactor(self.ROTATE_GEAR_RATIO * 60)
+
+        # TODO: get pid and feedforward values for arm and extension from sysid
+        # running the controller on the rio rather than on the motor controller
+        # to allow access to the velocity setpoint for feedforward
+        self.rotation_controller = ProfiledPIDController(
+            5, 0, 0, TrapezoidProfile.Constraints(maxVelocity=3, maxAcceleration=3)
+        )
+        # ignore 'meters', theres no generic one
+        self.rotation_simple_ff = SimpleMotorFeedforwardMeters(0.1, 0.5, 1)
+
+        # Create extension things
+        self.extension_motor = rev.CANSparkMax(
+            CanIds.Arm.extension, rev.CANSparkMax.MotorType.kBrushless
+        )
+        self.extension_motor.setIdleMode(rev.CANSparkMax.IdleMode.kCoast)
+        self.extension_encoder = self.extension_motor.getEncoder()
+        self.extension_encoder.setPositionConversionFactor(self.EXTEND_GEAR_RATIO)
+        self.extension_encoder.setVelocityConversionFactor(self.EXTEND_GEAR_RATIO * 60)
+        # assume retracted starting position
+        self.extension_encoder.setPosition(self.MIN_EXTENSION)
+        self.extension_controller = ProfiledPIDController(
+            1, 0, 0, TrapezoidProfile.Constraints(maxVelocity=2, maxAcceleration=5)
+        )
+        self.extension_simple_ff = SimpleMotorFeedforwardMeters(0.1, 0.5, 1)
+
+        self.absolute_encoder = DutyCycleEncoder(PwmPorts.arm_abs_encoder)
```

Comment:
Out of curiosity, is there a particular reason we're wiring this to the roboRIO rather than the motor controller?